config/deprecation.go

```diff
@@ -0,0 +1,362 @@
+	} else {
+		removalMajor, removalMinor = sinceMajor+1, 0
+		di.info.RemovalIn = fmt.Sprintf("%d.%d", removalMajor, removalMinor)
+	}
+
+	di.Level = None
+	if major > removalMajor || (major == removalMajor && minor >= removalMinor) {
+		di.Level = Error
+	} else if major > sinceMajor || (major == sinceMajor && minor >= sinceMinor) {
+		di.Level = Warn
+	}
+}
+
+// pluginDeprecationInfo holds all information about a deprecated plugin or it's options
+type pluginDeprecationInfo struct {
+	deprecationInfo
+
+	// Options deprecated for this plugin
+	Options []deprecationInfo
+}
+
+func (c *Config) incrementPluginDeprecations(category string) {
+	newcounts := []int64{1, 0}
+	if counts, found := c.Deprecations[category]; found {
+		newcounts = []int64{counts[0] + 1, counts[1]}
+	}
+	c.Deprecations[category] = newcounts
+}
+
+func (c *Config) incrementPluginOptionDeprecations(category string) {
+	newcounts := []int64{0, 1}
+	if counts, found := c.Deprecations[category]; found {
+		newcounts = []int64{counts[0], counts[1] + 1}
+	}
+	c.Deprecations[category] = newcounts
+}
+
+func (c *Config) collectDeprecationInfo(category, name string, plugin interface{}, all bool) pluginDeprecationInfo {
+	info := pluginDeprecationInfo{
+		deprecationInfo: deprecationInfo{
+			Name:  category + "." + name,
+			Level: None,
+		},
+	}
+
+	// First check if the whole plugin is deprecated
+	switch category {
+	case "inputs":
+		if pi, deprecated := inputs.Deprecations[name]; deprecated {
+			info.deprecationInfo.info = pi
+		}
+	case "outputs":
+		if pi, deprecated := outputs.Deprecations[name]; deprecated {
+			info.deprecationInfo.info = pi
+		}
+	}
+	info.determineEscalation(c.VersionMajor, c.VersionMinor)
+	if info.Level != None {
+		c.incrementPluginDeprecations(category)
+	}
+
+	// Check for deprecated options
+	walkPluginStruct(reflect.ValueOf(plugin), func(field reflect.StructField, value reflect.Value) {
+		// Try to report only those fields that are set
+		if !all && value.IsZero() {
+			return
+		}
+
+		tags := strings.SplitN(field.Tag.Get("deprecated"), ";", 3)
+		if len(tags) < 1 || tags[0] == "" {
+			return
+		}
+		optionInfo := deprecationInfo{Name: field.Name}
+		optionInfo.info.Since = tags[0]
+
+		if len(tags) > 1 {
+			optionInfo.info.Notice = tags[len(tags)-1]
+		}
+		if len(tags) > 2 {
+			optionInfo.info.RemovalIn = tags[1]
+		}
+		optionInfo.determineEscalation(c.VersionMajor, c.VersionMinor)
+
+		if optionInfo.Level != None {
+			c.incrementPluginOptionDeprecations(category)
+		}
+
+		// Get the toml field name
+		option := field.Tag.Get("toml")
+		if option != "" {
+			optionInfo.Name = option
+		}
+		info.Options = append(info.Options, optionInfo)
+	})
+
+	return info
+}
+
+func (c *Config) printUserDeprecation(category, name string, plugin interface{}) error {
+	info := c.collectDeprecationInfo(category, name, plugin, false)
+
+	switch info.Level {
+	case Warn:
+		prefix := "W! " + color.YellowString("DeprecationWarning")
```

Comment:
Isn't the colouring something should be handled by the logging handlers?

Reply:
Well but it doesn't. Should I remove the color? It made the warning/error more obvious during startup, but I agree that the general logging would benefit from colors...

Reply:
I would leave that to the general logging indeed, it is not a thing to fix/arrange in this PR.
Would be a nice added feature though!

Reply:
@Hipska will remove here. Do you care to open an issue?

Reply:
Well, it's not something I really *want*, just a nice to have. I don't care if it is there or not.

Reply:
Btw: We are directly using `log` here as the telegraf logging facility is not yet setup. `log` itself cannot do coloring... Any idea?

Reply:
I will keep the color as there is currently no way to introduce it otherwise due to direct usage of the `log` package.